src/zenml/integrations/modal/step_operators/modal_step_operator.py

```diff
@@ -0,0 +1,217 @@
+                build = BuildConfiguration(
+                    key=MODAL_STEP_OPERATOR_DOCKER_IMAGE_KEY,
+                    settings=step.config.docker_settings,
+                    step_name=step_name,
+                )
+                builds.append(build)
+
+        return builds
+
+    def launch(
+        self,
+        info: "StepRunInfo",
+        entrypoint_command: List[str],
+        environment: Dict[str, str],
+    ) -> None:
+        """Launch a step run on Modal.
+
+        Args:
+            info: The step run information.
+            entrypoint_command: The entrypoint command for the step.
+            environment: The environment variables for the step.
+
+        Raises:
+            subprocess.CalledProcessError: If the modal app exits with a non-zero code.
+            ValueError: If no Docker credentials are found for the container registry.
+        """
+        settings = cast(ModalStepOperatorSettings, self.get_settings(info))
+        image_name = info.get_image(key=MODAL_STEP_OPERATOR_DOCKER_IMAGE_KEY)
+        zc = Client()
+        stack = zc.active_stack
+
+        if not stack.container_registry:
+            raise ValueError(
+                "No Container registry found in the stack. "
+                "Please add a container registry and ensure "
+                "it is correctly configured."
+            )
+
+        if docker_creds := stack.container_registry.credentials:
+            docker_username, docker_password = docker_creds
+        else:
+            raise RuntimeError(
+                "No Docker credentials found for the container registry."
+            )
+
+        my_secret = modal.secret._Secret.from_dict(
+            {
+                "REGISTRY_USERNAME": docker_username,
+                "REGISTRY_PASSWORD": docker_password,
+            }
+        )
+
+        spec = modal.image.DockerfileSpec(
+            commands=[f"FROM {image_name}"], context_files={}
+        )
+
+        zenml_image = (
+            modal.Image._from_args(
+                dockerfile_function=lambda *_, **__: spec,
+                force_build=False,
+                image_registry_config=modal.image._ImageRegistryConfig(
+                    api_pb2.REGISTRY_AUTH_TYPE_STATIC_CREDS, my_secret
+                ),
+            )
+            .env(environment)
+            .run_commands(" ".join(entrypoint_command))
+        )
+
+        resource_settings = info.config.resource_settings
+        gpu_str = settings.gpu
+        if resource_settings.gpu_count is not None:
+            gpu_str += f":{resource_settings.gpu_count}"
+
+        app = modal.App(f"zenml-{info.run_name}-{info.step_run_id}")
+
+        async def run_sandbox():
+            with modal.enable_output():
+                async with app.run():
+                    await modal.Sandbox.create.aio(
+                        "bash",
+                        "-c",
+                        " ".join(entrypoint_command),
```

Comment:
Why do we need to specify the entrypoint both here as well as earlier when we define the image?

Reply:
That's a very good question for which I have no good answer. Maybe I'll test removing one or the other of them. Also maybe @kramstrom has an answer for that?

Reply:
You shouldn't need both, in the `run_commands` for the image you specify the commands to run when building the image (think docker's `RUN ...` statements)

On this line you specify the command to run in the sandbox

Reply:
Removing the command from when the image is created makes it fail. I'm guessing
that the command on the sandbox is actually redundant, but we need it because of
the way they create the Sandbox.

In any case, Modal team are reworking their API around Sandbox creation, so this
code will be updated in the coming weeks anyway, I think.

Reply:
@kramstrom how would I remove one of them, then? I tried removing the `run_commands` line from the image definition above, but then it broke / stopped working.

Reply:
what commands are you running in the entrypoint here?

Reply:
@kramstrom 'python -m zenml.entrypoints.entrypoint --entrypoint_config_source zenml.step_operators.step_operator_entrypoint_configuration.StepOperatorEntrypointConfiguration --deployment_id 86e4cbb8-cc5d-45b2-9969-e6956e200ee1 --step_name hello_world_on_modal --step_run_id b8f5ca23-2947-4463-be2c-2410655db816' something like this is the `entrypoint_command`... it's built at the same time as the image is built. so they should go together.